gensim/models/experimental/drmm_tks.py

```diff
@@ -0,0 +1,595 @@
+            shape : (num_samples, text_maxlen)
+        X2: the docs
+            shape : (num_samples, text_maxlen)
+        y: int {0, 1}
+            The relation between X1[i] and X2[j]
+            1 : X2[i] is relevant to X1[i]
+            0 : X2[i] is not relevant to X1[i]
+        """
+
+        num_samples = len(self.indexed_pair_list)
+        X1 = np.zeros((num_samples * 2, self.text_maxlen))
+        X2 = np.zeros((num_samples * 2, self.text_maxlen))
+        # To be uncommented when histogram support is included
+        # if self.hist_size is not None:
+        #     X2 = np.zeros((num_samples * 2, self.text_maxlen, self.hist_size))
+        y = np.zeros((num_samples * 2, 1))
+        X1[:] = self.pad_word_index
+        X2[:] = self.pad_word_index
+        y[::2] = 1
+        for i, (query, pos_doc, neg_doc) in enumerate(self.indexed_pair_list):
+            query_len = min(self.text_maxlen, len(query))
+            pos_doc_len = min(self.text_maxlen, len(pos_doc))
+            neg_doc_len = min(self.text_maxlen, len(neg_doc))
+
+            X1[i * 2, :query_len] = query[:query_len]
+            X2[i * 2, :pos_doc_len] = pos_doc[:pos_doc_len]
+            X1[i * 2 + 1, :query_len] = query[:query_len]
+            X2[i * 2 + 1, :neg_doc_len] = neg_doc[:neg_doc_len]
+        return X1, X2, y
+
+    def get_pair_list(self):
+        """Returns a list with query document pairs in the format
+        (query, positive_doc, negative_doc)
+
+        Example output:
+        -------------
+        [(q1, d+, d-), (q2, d+, d-), (q3, d+, d-), ..., (qn, d+, d-)]
+
+             where each query or document is a list of ints
+
+        Example:
+        -------
+        [(['When', 'was', 'Abraham', 'Lincoln', 'born', '?'],
+          ['He', 'was', 'born', 'in', '1809'],
+          ['Abraham', 'Lincoln', 'was', 'the', 'president',
+           'of', 'the', 'United', 'States', 'of', 'America']),
+
+         (['When', 'was', 'the', 'first', 'World', 'War', '?'],
+          ['It', 'was', 'fought', 'in', '1914'],
+          ['There', 'were', 'over', 'a', 'million', 'deaths']),
+
+         (['When', 'was', 'the', 'first', 'World', 'War', '?'],
+          ['It', 'was', 'fought', 'in', '1914'],
+          ['The', 'first', 'world', 'war', 'was', 'bad'])
+        ]
+
+        """
+        pair_list = []
+        for q, doc, label in zip(self.queries, self.docs, self.labels):
+            doc, label = (list(t)
+                          for t in zip(*sorted(zip(doc, label), reverse=True)))
+            for item in zip(doc, label):
+                if item[1] == 1:
+                    for new_item in zip(doc, label):
+                        if new_item[1] == 0:
+                            pair_list.append((q, item[0], new_item[0]))
+        return pair_list
+
+    def make_indexed_pair_list(self):
+        """Converts the existing word based pair list into an indexed format
+
+        Note: pair_list needs to be first created using get_pair_list"""
```

Comment:
link to method, see https://github.com/RaRe-Technologies/gensim/blob/fd6f6eb71996cd36a6236070f33efdfb37dde342/gensim/models/keyedvectors.py#L304